engine/src/main/scala/cromwell/webservice/routes/WesCromwellRouteSupport.scala

```diff
@@ -0,0 +1,90 @@
+import scala.concurrent.{ExecutionContext, TimeoutException}
+import scala.util.{Failure, Success}
+
+trait WesCromwellRouteSupport extends WebServiceUtils {
+
+  val workflowStoreActor: ActorRef
+
+  implicit val duration = ConfigFactory.load().as[FiniteDuration]("akka.http.server.request-timeout")
+  implicit val timeout: Timeout = duration
+
+  implicit def actorRefFactory: ActorRefFactory
+  implicit val materializer: ActorMaterializer
+  implicit val ec: ExecutionContext
+
+  def toResponse(workflowId: WorkflowId, workflowState: WorkflowState): WorkflowSubmitResponse = {
+    WorkflowSubmitResponse(workflowId.toString, workflowState.toString)
+  }
+
+  def submitRequest(formData: Multipart.FormData, isSingleSubmission: Boolean): Route = {
+
+    def getWorkflowState(workflowOnHold: Boolean): WorkflowState = {
+      if (workflowOnHold)
+        WorkflowOnHold
+      else WorkflowSubmitted
+    }
+
+    def sendToWorkflowStore(command: WorkflowStoreActor.WorkflowStoreActorSubmitCommand, warnings: Seq[String], workflowState: WorkflowState): Route = {
+      //sendToWorkflowStore
+      // NOTE: Do not blindly copy the akka-http -to- ask-actor pattern below without knowing the pros and cons.
+        onComplete(workflowStoreActor.ask(command).mapTo[WorkflowStoreSubmitActor.WorkflowStoreSubmitActorResponse]) {
+          case Success(w) =>
+            w match {
+              case WorkflowStoreSubmitActor.WorkflowSubmittedToStore(workflowId, _) =>
+                completeResponse(StatusCodes.Created, toResponse(workflowId, workflowState), warnings)
+              case WorkflowStoreSubmitActor.WorkflowsBatchSubmittedToStore(workflowIds, _) =>
+                completeResponse(StatusCodes.Created, workflowIds.toList.map(toResponse(_, workflowState)), warnings)
+              case WorkflowStoreSubmitActor.WorkflowSubmitFailed(throwable) =>
+                throwable.failRequest(StatusCodes.BadRequest, warnings)
+            }
+          case Failure(_: AskTimeoutException) if CromwellShutdown.shutdownInProgress() => CromwellApiService.serviceShuttingDownResponse
+          case Failure(e: TimeoutException) => e.failRequest(StatusCodes.ServiceUnavailable)
+          case Failure(e) => e.failRequest(StatusCodes.InternalServerError, warnings)
+        }
+    }
+
+    onComplete(materializeFormData(formData)) {
+      case Success(data) =>
+        PartialWorkflowSources.fromSubmitRoute(data, allowNoInputs = isSingleSubmission) match {
+          case Success(workflowSourceFiles) if isSingleSubmission && workflowSourceFiles.size == 1 =>
+            val warnings = workflowSourceFiles.flatMap(_.warnings)
+            sendToWorkflowStore(WorkflowStoreActor.SubmitWorkflow(workflowSourceFiles.head), warnings, getWorkflowState(workflowSourceFiles.head.workflowOnHold))
+          // Catches the case where someone has gone through the single submission endpoint w/ more than one workflow
+          case Success(workflowSourceFiles) if isSingleSubmission =>
```

Comment:
Hmm, so this `if` before `=>` thing enables multiple duplicate cases. I know this is pretty blunt, but I feel that this is a feature that Scala should not have and that we should not use. Rather, one case with an if/else if/else inside it. Of course, if this is a super idiomatic Scala thing, then I'll just get used to it.

(My reasoning: the middle `if` case is an `else if` without any explicit indication of the "else" part. If someone were to jump into the middle of a large blob of unfamiliar code without noticing an earlier duplicate case, they would "correctly" perceive the logic to be different than it actually was.)

Reply:
I found this link a while ago that helped me understand its benefits: https://stackoverflow.com/questions/58259575/what-are-advantages-of-pattern-matching-over-conditionals-in-fp-in-laymans-ter#:~:text=One%20of%20the%20most%20useful,never%20cause%20runtime%20exceptions%20%2F%20errors.

Reply:
Also, if you ever want to mob to discuss anything in here further, just let me know!